### PR TITLE
fixes #21657 - remove facets when host is removed

### PR DESCRIPTION
**Pull request by timogoebel**
Without this, you have to manually delete the facets from the database. There is no automatic cleanup.

cc: @ShimShtein

### CODE REVIEW

**Comment by timogoebel** [comment]
I see a lot of test failures like
```
PG::UndefinedTable: ERROR:  relation "test_classes" does not exist...
```

I believe, the tests aren't properly isolated. Will take a look.

**Comment by ShimShtein** [comment]
@jlsherrill , @thomasmckay Any objections? Could it cause issues to Katello?

**Comment by jlsherrill** [comment]
In katello we preferred this route, as deleting facets without deleting backend (candlepin or pulp) entities can lead to issues.  The idea being that deletion is directed through our foreman-task orchestration which properly removes those objects from the backend services before removing the facets and finally the host. 

One question though, how are you able to delete the host without deleting the facet first?  Are you missing a foreign key constraint on the host_id column on your facet?  It should prevent that.  

This change won't completely break katello (as i said the user ends up going through dynflow tasks when deleting a host), but may make it harder to detect if this breaks in the future.  Some sort of more official facet delete api may be nice, as i could see it useful for other things such as a puppet facet (where you may want to cleanup the puppet ca).

**Comment by sean797** [comment]
Do you think we should clear up any ~~fact~~ facets belonging to hosts that no long exist?

**Comment by ShimShtein** [comment]
> Do you think we should clear up any fact belonging to hosts that no long exist?

We already [do this](https://github.com/theforeman/foreman/blob/develop/app/models/host/base.rb#L17)

**Comment by sean797** [comment]
Sorry I mean facets. This will stop it happening anymore, but should there be a migration to clear any stale facets already in the database?

**Comment by ShimShtein** [comment]
@timogoebel, I think we should make it configurable - we can add another configuration option to `services/facets/entry.rb` and add `:dependent => destroy` accordingly.

**Comment by timogoebel** [comment]
@ShimShtein: Sounds like a good idea. Do you (as the "facet guy") by any chance want to take this over? :-)

**Comment by timogoebel** [comment]
@ShimShtein: Added a configuration parameter & tests.

**Comment by ShimShtein** [comment]
[test foreman]

**Comment by timogoebel** [comment]
Rebased to (hopefully) fix the tests.

**Review by boaz0 — APPROVED** [review]
LGTM 👍

**Comment by ShimShtein** [comment]
[test foreman]

**Comment by timogoebel** [comment]
Rebased.

**Comment by ShimShtein** [comment]
@timogoebel Tests are still red

**Comment by timogoebel** [comment]
*sights* pushed another attempt. I have a feeling, the configuration tests weren't isolated enough. I stubbed `Facets::ManagedHostExtensions.register_facet_relation` now. Hope, that helps.

**Comment by iNecas** [comment]
It probably didn't

**Comment by timogoebel** [comment]
Another attempt to fix the tests by reloading the host class following the suggestion by @ShimShtein.

**Comment by mmoll** [comment]
please rebase :)

**Comment by timogoebel** [comment]
Rebased.

**Comment by timogoebel** [comment]
[test foreman]

**Comment by ShimShtein** [comment]
[test foreman]

**Review by boaz0 — APPROVED** [review]
LGTM 💯

**Comment by tbrisker** [comment]
[test foreman]

**Comment by ShimShtein** [comment]
[test]

**Comment by timogoebel** [comment]
I actually have a feeling, the test failure could be related to the reloading of the host object. But I haven't had time to investigate.

**Comment by timogoebel** [comment]
I changed my opinion about the test failures being related. Rebased. Hopefully, tests are 🍏 now.

**Comment by mmoll** [comment]
@timogoebel please rebase :)

**Comment by timogoebel** [comment]
Rebased.

File: app/services/facets/entry.rb [hunk 1]
```diff
@@ -80,6 +80,14 @@ def template_compatibility_properties(*property_symbols)
       @compatibility_properties = property_symbols
     end
 
+    # Specify what should happen with the facet when the host object is deleted.
+    # [+:destroy+] The facet object will be deleted by calling the destroy record.
+    # [+:delete+] The facet will be deleted directly from the database without calling their destroy method.
+    def dependent(value = nil)
+      @dependent = value if value.present?
```
Comment (tbrisker) [inline app/services/facets/entry.rb]: This method looks like a combination of getter and setter, why not just use `attr_accessor`? Or name it in a way that has clear meaning, like `set_dependent_action`?

**Review by tbrisker — CHANGES_REQUESTED** [review]
I managed to reliably reproduce the test failures. They seem to be caused by the reloading of `host/managed.rb` in the facet test - commenting out those lines, or even just the one in `teardown` makes audit tests pass. Are you sure they are needed?
To reproduce locally: 
`SEED=30219 ruby -I.:test -e "ARGV.each{|f| require f}" test/unit/facet_test.rb test/models/concerns/audit_extensions_test.rb`

**Comment by timogoebel** [comment]
@tbrisker: Thanks for taking a look. :-)

> Are you sure they are needed?

I just pushed a version with these lines commented out. Let's see what that breaks.

**Comment by ShimShtein** [comment]
@tbrisker facets pollute the `Host::Managed` class with new relations. Those relations cannot be undone, so if you are adding a facet for test only, you end up with unwanted relation methods for all other tests. This is what started this whole mess.

**Comment by timogoebel** [comment]
Tests passed without the reload code. I just pushed a version that renames the DSL to use `set_dependent_action`. I hope, we're good to go now. *sights*

**Comment by timogoebel** [comment]
@tbrisker, @ShimShtein: Tests are green now. 🌲

**Review by tbrisker — DISMISSED** [review]
Works for me, any more comments @ShimShtein ?

File: test/unit/facet_test.rb [hunk 2]
```diff
@@ -14,12 +14,6 @@ class FacetTest < ActiveSupport::TestCase
     Facets.stubs(:configuration).returns(@config)
   end
 
-  teardown do
```
Comment (ShimShtein) [inline test/unit/facet_test.rb]: Without those lines, cloning `Host::Managed` will try to clone `test_facet`. I don't think it's desirable.

After this test Host::Managed will contain all properties related to `test_facet` and `namespaced_facet`

**Comment by timogoebel** [comment]
I re-added the teardown lines for the tests.

**Comment by tbrisker** [comment]
@timogoebel I think you forgot to push, and I think it will still fail (it did when i was testing on the older version). Perhaps we need to find some other way to remove the association that doesn't include reloading the model.

**Comment by timogoebel** [comment]
> Perhaps we need to find some other way to remove the association that doesn't include reloading the model.

To be honest, I'm out of ideas. I could try to not add the association in the first place by stubbing it, but otherwise: I have no idea how to proceed here. Suggestions are very welcome.

I think I did push my changes.
@tbrisker: What is missing?

**Comment by ShimShtein** [comment]
I am not sure it's a good one, but maybe we can use an inherited throwaway class here?
Something along the lines:

``` ruby
class FacetTest < ActiveSupport::TestCase
  class TestHost < Host::Managed
  end
end
```
I am not sure it will work, and what ActiveRecord trickery we will need to invoke, but at least it will not pollute the original `Host::Managed` class.

**Comment by leewaa** [comment]
Nothing else is missing here ? Looks ok to me and tests are green.

No other objections to this one ?

Just needs a rebase

**Comment by timogoebel** [comment]
Rebased.

Can we get this in? I don't think rewriting the way we test facets is actually within the scope of this issue. We should probably do it at some time, but we can do that later I believe.

**Review by mmoll — APPROVED** [review]

**Review by tbrisker — APPROVED** [review]
LGTM, Thanks @timogoebel and everyone involved in reviewing :)

**Comment by timogoebel** [comment]
Thanks, guys.

I updated the documentation to include this: http://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Plugin#section-26

**Pull request merged**